docs/configuration/http_filters/lua_filter.rst

```diff
@@ -0,0 +1,325 @@
+      {[":status"] = "403",
+       ["upstream_foo"] = headers["foo"]},
+      "nope")
+  end
+
+Stream handle API
+-----------------
+
+When Envoy loads the script in the configuration, it looks for two global functions that the
+script defines:
+
+.. code-block:: lua
+
+  function envoy_on_request(request_handle)
+  end
+
+  function envoy_on_response(response_handle)
+  end
+
+A script can define either or both of these functions. During the request path, Envoy will
+run *envoy_on_request* as a coroutine, passing an API handle. During the response path, Envoy will
+run *envoy_on_response* as a coroutine, passing an API handle.
+
+.. attention::
+
+  It is critical that all interaction with Envoy occur through the passed stream handle. The stream
+  handle should not be assigned to any global variable and should not be used outside of the
+  coroutine. Envoy will fail your script if the handle is used incorrectly.
+
+The following methods on the stream handle are supported:
+
+headers()
+^^^^^^^^^
+
+.. code-block:: lua
+
+  headers = handle:headers()
+
+Returns the stream's headers. The headers can be modified as long as they have not been sent to
+the next filter in the header chain. For example, they can be modified after an *httpCall()* or
+after a *body()* call returns. The script will fail if the headers are modified in any other
+situation.
+
+Returns a :ref:`header object <config_http_filters_lua_header_wrapper>`.
+
+body()
+^^^^^^
+
+.. code-block:: lua
+
+  body = handle:body()
+
+Returns the stream's body. This call will cause Envoy to yield the script until the entire body
+has been buffered. Note that all buffering must adhere to the flow control policies in place.
+Envoy will not buffer more data than is allowed by the connection manager.
+
+Returns a :ref:`buffer object <config_http_filters_lua_buffer_wrapper>`.
+
+bodyChunks()
+^^^^^^^^^^^^
+
+.. code-block:: lua
+
+  iterator = handle:bodyChunks()
+
+Returns an iterator that can be used to iterate through all received body chunks as they arrive.
+Envoy will yield rthe script in between chunks, but *will not buffer* them. This can be used by
+a script to inspect data as it is streaming by.
+
+.. code-block:: lua
+
+  for chunk in request_handle:bodyChunks() do
+    request_handle:log(0, chunk:byteSize())
+  end
+
+Each chunk the iterator returns is a :ref:`buffer object <config_http_filters_lua_buffer_wrapper>`.
+
+trailers()
+^^^^^^^^^^
+
+.. code-block:: lua
+
+  trailers = handle:trailers()
+
+Returns the stream's trailers. May return nil if there are no trailers. The trailers may be
+modified before they are sent to the next filter.
+
+Returns a :ref:`header object <config_http_filters_lua_header_wrapper>`.
+
+log()
+^^^^^
+
+.. code-block:: lua
+
+  handle:log(level, message)
+
+Logs a message using Envoy's application logging. *level* is an integer in the range 0 - 5 which
+maps to the spdlog levels *trace* - *critical*. *message* is a string to log.
+
+httpCall()
```

Comment:
It will be awesome when we can do gRPC here BTW, but I know this is not for MVP.

Reply:
I honestly have no idea how this would work. The issue is protos, not making the call. Definitely not in the MVP. I will put it in as a potential follow up.

Reply:
I have started crying :( .... gRPC in lua...

Reply:
Yeah, I know the protos are the issue. There are libraries out there, e.g. https://github.com/urbanairship/protobuf-lua, but NFI how close to being useful for general purpose gRPC in a filter they will be.